src/axolotl/utils/config/models/input/v0_4_1/__init__.py

```diff
@@ -1548,3 +1550,15 @@ def check_multigpu_unsloth(cls, data):
                     "unsloth_lora_mlp, unsloth_lora_qkv, and unsloth_lora_o are not compatible with multi-GPU training."
                 )
         return data
+
+    @model_validator(mode="before")
+    @classmethod
+    def check_adopt_torch_version(cls, data):
+        if (data.get("optimizer") is not None) and ("adopt" in data.get("optimizer")):
+            env_capabilities = data.get("env_capabilities")
+            torch_version = env_capabilities.get("torch_version")
```

Comment:
if env_capabilities is `Optional` this will fail with a KeyError or AttributeError I believe

Reply:
in load_cfg I've set `env_capabilities` to include the torch_version, so during regular usage it will never be `None`. I previously set `env_capabilities` to be `Optional` so that it passes some of the tests. For here I think I can make it such that if `env_capabilities` is `None` it will perform a check for torch version

Reply:
Right, but if in regular usage it should not be optional, we should remove the optional typehint. Otherwise you need to add guards here for when it is passed as none otherwise it will fail spectacularly and someone will ultimately end up filing a bug report